homeassistant/components/saj/sensor.py

```diff
@@ -0,0 +1,191 @@
+"""SAJ solar inverter interface."""
+import asyncio
+from datetime import date, timedelta
+import logging
+
+import pysaj
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_HOST,
+    DEVICE_CLASS_POWER,
+    DEVICE_CLASS_TEMPERATURE,
+    POWER_WATT,
+    TEMP_CELSIUS,
+    TEMP_FAHRENHEIT,
+)
+from homeassistant.core import callback, CALLBACK_TYPE
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_point_in_utc_time
+from homeassistant.loader import bind_hass
+from homeassistant.util import dt as dt_util
+
+_LOGGER = logging.getLogger(__name__)
+
+INTERVAL = timedelta(seconds=5)
+INTERVAL_BACKOFF_1 = timedelta(minutes=1)
+INTERVAL_BACKOFF_2 = timedelta(minutes=2)
+INTERVAL_BACKOFF_3 = timedelta(minutes=5)
+
+PLATFORM_SCHEMA = vol.All(
+    PLATFORM_SCHEMA.extend(
+        {vol.Required(CONF_HOST): cv.string}, extra=vol.PREVENT_EXTRA
+    )
+)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up SAJ sensors."""
+
+    # Init all sensors
+    sensor_def = pysaj.Sensors()
+
+    # Use all sensors by default
+    hass_sensors = []
+
+    for sensor in sensor_def:
+        hass_sensors.append(SAJsensor(sensor))
+
+    saj = pysaj.SAJ(config[CONF_HOST])
+
+    async_add_entities(hass_sensors)
+
+    intervals = (INTERVAL, INTERVAL_BACKOFF_1, INTERVAL_BACKOFF_2, INTERVAL_BACKOFF_3)
+
+    async def async_saj(event):
+        """Update all the SAJ sensors."""
+        tasks = []
+
+        values = await saj.read(sensor_def)
+
+        for sensor in hass_sensors:
+            state_unknown = False
+            if not values:
+                # SAJ inverters are powered by DC via solar panels and thus are
+                # offline after the sun has set. If a sensor resets on a daily
+                # basis like "today_yield", this reset won't happen automatically.
+                # Code below checks if today > day when sensor was last updated
+                # and if so: set state to None.
+                # Sensors with live values like "temperature" or "current_power"
+                # will also be reset to None.
+                if (sensor.per_day_basis and date.today() > sensor.date_updated) or (
+                    not sensor.per_day_basis and not sensor.per_total_basis
+                ):
+                    state_unknown = True
+            task = sensor.async_update_values(unknown_state=state_unknown)
+            if task:
+                tasks.append(task)
+        if tasks:
+            await asyncio.wait(tasks)
+        return values
+
+    async_track_time_interval_backoff(hass, async_saj, intervals)
+
+    return True
```

Comment:
Nothing is checking this return value.

Reply:
Indeed, my bad, probably a leftover from testing some things.
Removed the statement.